#Перевірка статиcтичних гіпотез. Z-test.



0. Зчитайте дані з `data.csv` у змінну data, яка має тип pandas.DataFrame. Ми будемо далі працювати з цією змінною.

In [9]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats

# Налаштування для візуалізації
plt.rcParams['figure.figsize'] = (12, 7)
plt.rcParams['font.size'] = 12

In [10]:
# Зчитування даних
data = pd.read_csv('data.csv')

# Перевірка
print(type(data))
print(data.head())  # показати перші 5 рядків

<class 'pandas.core.frame.DataFrame'>
   Unnamed: 0       data
0           0  17.499453
1           1  19.662399
2           2   7.182823
3           3  29.841625
4           4   9.239386


Запустіть код нижче. Для коректної робити всіх подальших методів дані мають бути у вигляді одновимірного `numpy.ndarray` вектора та мати розмірність (100,). Така розмірність означає, що дані - одновимірні (колонка), якщо б розмірність була (100, 1), то дані сприймаються методами як двовимірні (таблиця), хоча для нас виглядати можуть так само.

In [11]:
data = data.data.values
data.shape

(100,)

**Завдання 1**. Зчитані дані - це сума покупок в доларах наших покупців на сайті протягом липня 2025 року.

До цього місяця, ми мали продажі в середньому на чек $20.

Необхідно зʼясувати, чи вийшло у нас статистично значущо **підвищити** середній чек за липень 2025?

Аби дати відповідь - ми проведемо z-test на рівні значущості $\alpha=0.05$ двома способами. В цьому завданні ми будемо виконувати обчислення "вручну" з використанням бібліотек numpy та scipy.stats подібно до прикладу в лекції. Для цього виконайте наступні 7 кроків. Правильне виконання кожного з кроків оцінюється в 1 бал.

1. Запишіть параметри задачі у змінні Python:
    - `sample_mean` - $\bar{x}$, середнє значення в вибірці
    - `population_mean` - $\mu_0$, середнє значення в популяції (тобто те, з яким ми порівнюємо середнє вибіркове значення)
    - `population_std` - $\sigma$, вибіркове стандартне відхилення, яке ми вважаємо, що є рівним ст. відх. популяції, адже маємо достатньо велику вибірку.
    - `sample_size` - $n$, розмір вибірки
    - `alpha` - $\alpha$ рівень значущості
    
    **Увага!** Для обчислення стандартного відхилення маємо скористатись функцією `np.std(your_dataframe, ddof=1)`. Чому так - розбираємо в лекції про t-test.


2. Визначте для цієї задачі:
    - якою є гіпотеза $H_0$
    - якою є альтернативна гіпотеза $H_a$
    - з яким типом тесту ми маємо справу - лівосторонній, правосторонній чи двосторонній.

3. Обчисліть стандартну помилку SE.
4. Розрахуйте z-статистику.
5. Знайдіть критичне z-значення з допомогою бібліотеки stats.
6. Обчисліть p-value з допомогою бібліотеки stats.
7. Прийміть рішення, чи відхиляєте ви гіпотезу $H_0$. Для прийняття рішення зробіть і порівняння z-статистики з критичним значенням, і проаналізуйте p-value.

🔬 Частина 1: Розрахунок вручну
*1. Запишемо параметри задачі у змінні Python:*


In [18]:
# 1. Параметри задачі
sample_mean = np.mean(sample)                     # x̄ Середній чек
population_mean = 20                              # µ0
population_std = np.std(sample, ddof=1)          # σ Стандартне відхилення
sample_size = len(sample)                         # n 
alpha = 0.05                                     # рівень значущості

In [17]:
# Вибірка
sample = data  # data вже є одновимірний масив

*2: Формулюємо гіпотези*

**Гіпотези**   

H₀: μ = 20 (середній чек у липні = 20 дол.)  
Hа: μ > 20 (середній чек у липні > 20 дол.)  
Тип тесту: правосторонній (ми перевіряємо, чи збільшвся чек)  

**Чому саме правосторонній тест?**

Ми перевіряємо **чи збільшився чек** (не просто став інакшм), тому ми перевіряємо, чи середній чек **збільшився**, а не просто змінився.


In [21]:
print("Середнє вибірки:", sample_mean)
print("Стандартне відхилення:", population_std)
print("Розмір вибірки:", sample_size)

Середнє вибірки: 19.37755319670441
Стандартне відхилення: 6.25356123240434
Розмір вибірки: 100


*3: Обчислимо стандартну помилку SE.*

In [22]:
SE = population_std / np.sqrt(sample_size)
print("Стандартна помилка:", SE)

Стандартна помилка: 0.625356123240434


*4: Розраховуємо z-статистику.*

In [24]:
z_stat = (sample_mean - population_mean) / SE
print("Z-статистика:", z_stat)

Z-статистика: -0.995347738933505


*5. Знаходимо критичне z-значення з допомогою бібліотеки stats.*

In [26]:
z_critical = stats.norm.ppf(1 - alpha)  # правосторонній тест
print("Критичне z-значення:", z_critical)

Критичне z-значення: 1.6448536269514722


*6. Обчислюємо p-value з допомогою бібліотеки stats.*

In [28]:
p_value = 1 - stats.norm.cdf(z_stat)  # правосторонній тест
print("P-value:", p_value)

P-value: 0.8402164165464012


*7. Прийняття рішення*

In [30]:
if z_stat > z_critical:
    print("Відхиляємо H0 — середній чек статистично значущий вищий за $20")
else:
    print("Не відхиляємо H0 — немає статистично значущого збільшення середнього чека")

if p_value < alpha:
    print("P-value: Відхиляємо H0")
else:
    print("P-value: Не відхиляємо H0")

Не відхиляємо H0 — немає статистично значущого збільшення середнього чека
P-value: Не відхиляємо H0


✅*Висновки:*
Z-статистика: -0.995 < 1.645 (критичне значення)
P-value: 0.840 > α = 0.05
За порівнянням Z-статистики з критичним значенням: Z менше критичного, отже не відхиляємо H0.
За P-value: P-value значно більше 0.05, отже теж не відхиляємо H0.
Пояснення: Немає статистично значущого збільшення середнього чеку порівняно зі значенням $20. Фактичне середнє в липні не перевищує середнє минулих періодів у значимий спосіб.

**Завдання 2.** Виконайте обчислення z-test з використанням бібліотеки statsmodels.

Отримайте z-статистику та р-значення.

Виведіть p-значення та зробіть висновок, чи ми маємо достатньо доказів, аби стверджувати, що середній чек зріс.

Чи зійшлись значення z-статистику та р-значення в цьому завданні з попередніми обчисленнями?

Завдання оцінюється в 3 бали.

In [31]:
from statsmodels.stats import weightstats as stests

# Виконання z-тесту
z_stat_sm, p_value_sm = stests.ztest(sample, value=20, alternative='larger')  # правосторонній тест

print("Z-статистика (statsmodels):", z_stat_sm)
print("P-value (statsmodels):", p_value_sm)

# Висновок
alpha = 0.05
if p_value_sm < alpha:
    print("Відхиляємо H0 — є статистично значуще збільшення середнього чеку")
else:
    print("Не відхиляємо H0 — недостатньо доказів, щоб стверджувати про збільшення середнього чеку")


Z-статистика (statsmodels): -0.9953477389335053
P-value (statsmodels): 0.8402164165464012
Не відхиляємо H0 — недостатньо доказів, щоб стверджувати про збільшення середнього чеку


✅ Результати  z-статистику та р-значення в цьому завданні з попередніми обчисленнями (ручним) зійшлись